soc/riscv/riscv-privilege/telink_w91/blocking_core/blocking.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2024 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <compiler.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/atomic.h>
+#include <ipc/ipc_based_driver.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(blocking_core_w91);
+
+enum {
+	IPC_DISPATCHER_BLOCKING_SET_STATE_ADDR = IPC_DISPATCHER_BLOCKING,
+	IPC_DISPATCHER_BLOCKING_STOP_CORE_REQ,
+};
+
+enum {
+	BLOCKING_INVALID_STATE,
+	BLOCKING_INITIATED_STATE,
+	BLOCKING_CORE_ACTIVE_STATE,
+	BLOCKING_CORE_STOP_REQ_STATE,
+	BLOCKING_CORE_STOPPED_STATE,
+	BLOCKING_CORE_ACTIVE_REQ_STATE,
+};
+
+static atomic_t blocking_state = BLOCKING_INVALID_STATE;
+static struct ipc_based_driver ipc_data;    /* ipc driver data part */
+
+/* APIs implementation: set address of blocking state */
+static size_t pack_blocking_w91_set_state_addr(uint8_t inst, void *unpack_data, uint8_t *pack_data)
+{
+	uint32_t *p_set_state_addr_req = unpack_data;
+	size_t pack_data_len = sizeof(uint32_t) + sizeof(uint32_t);
+
+	if (pack_data != NULL) {
+		uint32_t id = IPC_DISPATCHER_MK_ID(IPC_DISPATCHER_BLOCKING_SET_STATE_ADDR, inst);
+
+		IPC_DISPATCHER_PACK_FIELD(pack_data, id);
+		IPC_DISPATCHER_PACK_FIELD(pack_data, *p_set_state_addr_req);
+	}
+
+	return pack_data_len;
+}
+
+IPC_DISPATCHER_UNPACK_FUNC_ONLY_WITH_ERROR_PARAM(blocking_w91_set_state_addr);
+
+static int blocking_w91_set_state_addr(uint32_t addr)
+{
+	int err;
+
+	IPC_DISPATCHER_HOST_SEND_DATA(&ipc_data, 0,
+			blocking_w91_set_state_addr, &addr, &err,
+			CONFIG_BLOCKING_CORE_TELINK_W91_IPC_RESPONSE_TIMEOUT_MS);
+
+	return err;
+}
+
+/* APIs implementation: core stop request event */
+static _attribute_ram_code_ void blocking_w91_stop_core(void)
+{
+	unsigned int key;
+
+	/* Convert the check state period to approximate number of "nop" instructions */
+	uint32_t nop_count = CONFIG_BLOCKING_CORE_TELINK_W91_CHECK_STATE_PERIOD_US *
+			(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / USEC_PER_SEC);
+
+	uint32_t steps_timeout = CONFIG_BLOCKING_CORE_TELINK_W91_CORE_STOP_TIMEOUT_US /
+			CONFIG_BLOCKING_CORE_TELINK_W91_CHECK_STATE_PERIOD_US;
+	uint32_t steps_count = 0;
+
+	key = irq_lock();
+
+	atomic_set(&blocking_state, BLOCKING_CORE_STOPPED_STATE);
+
+	while (atomic_get(&blocking_state) != BLOCKING_CORE_ACTIVE_REQ_STATE) {
```

Comment:
Seems we left this cycle by timeout always...

Reply:
The N22 core sets BLOCKING_CORE_ACTIVE_REQ_STATE